DM-45237: Add ability to disable file validation info on ingest-files and import

timj opened the pull request:
## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`
- [ ] (if changing dimensions.yaml) make a copy of dimensions.yaml in `configs/old_dimensions`

timj committed:
Allow butler ingest-files to disable file attribute tracking

timj committed:
Add ability to disable file validation tracking on butler import

timj committed:
Remove file existence checking on ingest

* Existence checking has a large overhead for remote URIs and
  can make an ingest untenable with tens of thousands of files.
* Existence would eventually be checked for any transfer mode
  other than None or direct.
* Existence would eventually be checked if file sizes are being
  recorded (although it is likely that anyone using ingest
  would turn that off to save time).

On balance, removing the check entirely seems to be the
best approach rather than only doing it for local URIs
or only doing it if "direct and no track-file-attrs"

timj commented on python/lsst/daf/butler/datastores/fileDatastore.py:
```diff
@@ -963,12 +961,6 @@ def _standardizeIngestPath(
-        if not srcUri.exists():
-            raise FileNotFoundError(
-                f"Resource at {srcUri} does not exist; note that paths to ingest "
-                f"are assumed to be relative to {self.root} unless they are absolute."
```
I can't see anyway around removing this. Maybe I should do the check if the ingest is direct or None and then only if local file? There is no reason to do it for a copy since the copy will fail later on anyhow.

ktlim replied:
This doesn't seem like a primary function of this method, given its name and description.  I'm fine with a check (if any) happening elsewhere.

timj replied:
It sounds like any check is going to be prohibitively slow. Options seem to be:

* Never check and hope the file really is there when you ingest with direct / no-track-file-attrs
* Always check if local else never check
* Only check if there are fewer than 200 files being ingested
* Always check the first 10 (for example) regardless and if those do exist trust that the person has a reasonable set of data that is likely to really exist.

I think I'm talking myself into the last option (maybe above this method). That would catch the egregious case where someone is trying to ingest without really having access to the remote bucket or they copied the export file to their laptop and are assuming the files are still accessible. Even only checking the first might be enough.

cc/ @TallJimbo in case you have an opinion.

TallJimbo replied:
This is all for `direct` and `None` transfer modes, and not the others, right?  Not some case where a genuine artifact transfer silently fails?  If that's the case then some sort of spot-checking like the last bullet sounds reasonable.  Maybe the first one and a random 1% or something thereafter?

timj replied:
Right. If "copy" mode is being used then the ingest will break as soon as it tries to copy the file (maybe "symlink" will also fail?). If we are tracking file attributes the ingest will fail when it can't find the size.

So we only really need to do any spot checks if we are using direct/None with no-track-file-attrs.

timj replied:
Ok. I've changed it to only check 200 of the files being ingested. That doesn't take very long and you can at least get a sense that things are probably okay.

timj committed:
Remove pytest-openfiles

Closes #802

timj committed:
Add news fragments

timj committed:
Cap the number of checks for file existence during ingest

We want to check some of the files to ensure that they are
accessible at all but we don't want to check every file.
If there are thousands of files on a remote bucket this can take
a long time.

timj commented on python/lsst/daf/butler/datastores/fileDatastore.py:
```diff
@@ -1122,7 +1127,24 @@ def _extractIngestInfo(
+        if n_datasets <= max_checks:
+            check_every_n = 1
+        elif transfer in ("direct", None):
+            check_every_n = int(n_datasets / max_checks + 1)  # +1 so that if n < 100 the answer is 1.
```
@ktlim I've asked for a re-review since this change is quite significant. The code coverage doesn't cover the important bit because none of the tests try to ingest large numbers of files. I have verified that @gpdf 's example does only check 200 files out of 155k and that doesn't notable affect the import time.

timj committed:
Fix new mypy warning with dataclasses

python/lsst/daf/butler/formatters/json.py:136: note: Possible overload variants:
python/lsst/daf/butler/formatters/json.py:136: note:     def asdict(obj: DataclassInstance) -> dict[str, Any]
python/lsst/daf/butler/formatters/json.py:136: note:     def [_T] asdict(obj: DataclassInstance, *, dict_factory: Callable[[list[tuple[str, Any]]], _T]) -> _T

timj committed:
Fix type issue found by mypy 1.11

We use a default_key=None in the log handling code to imply
the root logger. Changing the types in split_kv to use str | None
fixes the problem. The --log-levels option wants to use None
to indicate the default logger should be affected and saves
"" to mean the root logger (the default logger is "lsst").

timj committed:
Change __exit__ return value

New typeshed now says return type is bool | None.

timj committed:
Ignore a mypy 1.11 warning for now

python/lsst/daf/butler/cli/utils.py:793: error: Argument "cls" to "option" has incompatible type "partial[MWOption]"; expected "type[Option] | None"  [arg-type]

timj committed:
Ignore mypy 1.11 warnings in formatter

The suggested fix on https://github.com/python/mypy/issues/17550 causes
another problem that we ignore for now.

timj commented on python/lsst/daf/butler/cli/utils.py:
```diff
@@ -352,10 +352,10 @@ def split_kv(
     unseparated_okay: bool = False,
     return_type: type[dict] | type[tuple] = dict,
-    default_key: str = "",
+    default_key: str | None = "",
```
This `None` is here specifically to support `--log-levels` because the code in `cliLog` is using `None` to indicate the default logger and "" or "." to indicate the root logger. They are not the same because we want `--log-level INFO` to affect the `lsst` logger and not the root logger.

timj commented on python/lsst/daf/butler/cli/utils.py:
```diff
@@ -789,7 +790,7 @@ class MWOptionDecorator:
 
     def __init__(self, *param_decls: Any, **kwargs: Any) -> None:
-        self.partialOpt = partial(click.option, *param_decls, cls=partial(MWOption), **kwargs)
+        self.partialOpt = partial(click.option, *param_decls, cls=partial(MWOption), **kwargs)  # type: ignore
```
mypy doesn't think that `partial(MWOption)` is the same as an option and so now complains.

andy-slac commented on doc/changes/DM-45237.misc.rst:
```diff
@@ -0,0 +1,5 @@
+File ingest no longer checks that ever file exists.
```
```suggestion
File ingest no longer checks that every file exists.
```

andy-slac commented on python/lsst/daf/butler/datastores/fileDatastore.py:
```diff
@@ -1122,7 +1127,24 @@ def _extractIngestInfo(
         filtered = []
-        for dataset in datasets:
+
+        # Ingest could be given tens of thousands of files. It is not efficient
+        # to check for the existence of every single file (especially if they
+        # are remote URIs) but in some transfer modes the files will be checked
+        # anyhow when they are relocated. For direct or None transfer modes
+        # it is possible to not know if the file is accessible at all.
+        # Therefore limit number of files that will be checked (but always
+        # include the first one).
```
Could there be a use case when clients want to check every file even if there are thousands of them? 200 looks like an arbitrary number, maybe we want a new option for that (or per-datastore config parameter)?

timj replied:
Yes. It's an issue. It's why part of the conversation was considering always doing the checks for POSIX and only skipping for remote. Gregory's original 150k import example was so long (many hours) that it seemed like it wouldn't ever be something people wanted to do. I picked 200 because that's about the size of the focal plane... and it doesn't take very long.

andy-slac reviewed (approved):
Looks great.
